gateway/session_manager.go

```diff
@@ -129,54 +209,26 @@ func (l *SessionLimiter) ForwardMessage(r *http.Request, currentSession *user.Se
 			rateScope = allowanceScope + "-"
 		}
 		if globalConf.EnableSentinelRateLimiter {
-			rateLimiterKey := RateLimitKeyPrefix + rateScope + currentSession.KeyHash()
-			rateLimiterSentinelKey := RateLimitKeyPrefix + rateScope + currentSession.KeyHash() + ".BLOCKED"
-
-			go l.doRollingWindowWrite(key, rateLimiterKey, rateLimiterSentinelKey, currentSession, store, globalConf, apiLimit, dryRun)
-
-			// Check sentinel
-			_, sentinelActive := store.GetRawKey(rateLimiterSentinelKey)
-			if sentinelActive == nil {
-				// Sentinel is set, fail
+			if l.limitSentinel(currentSession, key, rateScope, store, globalConf, apiLimit, dryRun) {
 				return sessionFailRateLimit
 			}
 		} else if globalConf.EnableRedisRollingLimiter {
-			rateLimiterKey := RateLimitKeyPrefix + rateScope + currentSession.KeyHash()
-			rateLimiterSentinelKey := RateLimitKeyPrefix + rateScope + currentSession.KeyHash() + ".BLOCKED"
-
-			if l.doRollingWindowWrite(key, rateLimiterKey, rateLimiterSentinelKey, currentSession, store, globalConf, apiLimit, dryRun) {
+			if l.limitRedis(currentSession, key, rateScope, store, globalConf, apiLimit, dryRun) {
 				return sessionFailRateLimit
 			}
 		} else {
-			// In-memory limiter
-			if l.bucketStore == nil {
-				l.bucketStore = memorycache.New()
-			}
-
-			bucketKey := key + ":" + rateScope + currentSession.LastUpdated
-			currRate := apiLimit.Rate
-			per := apiLimit.Per
-
-			// DRL will always overflow with more servers on low rates
-			rate := uint(currRate * float64(DRLManager.RequestTokenValue))
-			if rate < uint(DRLManager.CurrentTokenValue()) {
-				rate = uint(DRLManager.CurrentTokenValue())
+			var n float64
+			if DRLManager.Servers != nil {
+				n = float64(DRLManager.Servers.Count())
 			}
-
-			userBucket, err := l.bucketStore.Create(bucketKey, rate, time.Duration(per)*time.Second)
-			if err != nil {
-				log.Error("Failed to create bucket!")
-				return sessionFailRateLimit
-			}
-
-			if dryRun {
-				// if userBucket is empty and not expired.
-				if userBucket.Remaining() == 0 && time.Now().Before(userBucket.Reset()) {
+			if n <= 1 || n > apiLimit.Rate {
```

Comment:
It should be more complex than this. 
We need to introduce constant, which is configurable via the config. 
Additionally, it is not only about `Rate`, but it also depends on `Per`. We need to have Request per second in this formula. So formula will be `N*C > (apiLimit.Rate/apiLimit.Per)`. And C by default will be smth like 5.